Account Login and Verification Support

CnCRAZER opened the pull request:
### Establishes 2 new flags within `ClientDefinitions.ini`:
- `UseCnCNetAPI`
- `CnCNetAPIUrl`

Setting `UseCnCNetAPI` to `Yes` or `True` allows online login window to connect to cncnet ladder account and retrieve active nicknames. Setting this flag to `No` or `False` hides this field entirely.


<img width="1922" height="1112" alt="image" src="https://github.com/user-attachments/assets/ed602ac8-e743-4d7a-8188-46fd75a2f7ce" />

<img width="1922" height="1112" alt="image" src="https://github.com/user-attachments/assets/87016aaa-9ebd-4968-9eff-0ef06007077d" />

<img width="1922" height="1112" alt="image" src="https://github.com/user-attachments/assets/2c6dde67-a096-4ae3-be24-ecf2b1ec73d3" />

### Verified Account Only Option

<img width="686" height="393" alt="image" src="https://github.com/user-attachments/assets/5ecd8685-2afb-44cf-815d-7c8853a817d2" />

CnCRAZER committed:
Add CnCNet API-based account login and verification

Introduces new UI windows and backend logic for CnCNet account login, nickname management, and user verification via the CnCNet API.

Player list now displays verified status, and configuration options for API usage and endpoint are added.

CnCRAZER committed:
Update CnCNetAPI.cs

CnCRAZER committed:
Update CnCNetAPI.cs

CnCRAZER committed:
House Keeping

CnCRAZER committed:
Add verified account-only option for CnCNet games

Introduces a 'verified accounts only' setting for CnCNet multiplayer game creation and lobbies.

Updates UI to allow hosts to restrict rooms to verified CnCNet accounts, enforces restriction on join and in the lobby, and propagates the setting through game creation and broadcast logic.

CnCRAZER committed:
Fix multi-connection crashes

CnCRAZER committed:
Relocate verified accounts only checkbox

SadPencil commented on DXMainClient/DXGUI/Multiplayer/CnCNet/CnCNetLobby.cs:
```diff
@@ -1506,7 +1691,7 @@ private void GameBroadcastChannel_CTCPReceived(object sender, ChannelCTCPEventAr
             string[] splitMessage = msg.Split(new char[] { ';' });
 
-            if (splitMessage.Length != 12)
+            if (splitMessage.Length < 12)
```
huh?

CnCRAZER commented on ClientCore/ClientConfiguration.cs:
```diff
@@ -389,6 +389,13 @@ public string GetGameExecutableName()
 
         public bool DisplayPlayerCountInTopBar => clientDefinitionsIni.GetBooleanValue(SETTINGS, "DisplayPlayerCountInTopBar", false);
 
+        public bool UseCnCNetAPI => clientDefinitionsIni.GetBooleanValue(SETTINGS, "UseCnCNetAPI", true);
+
+        /// <summary>
+        /// Base URL for the CnCNet API, typically ends with /api/v1/.
+        /// </summary>
+        public string CnCNetApiUrl => clientDefinitionsIni.GetStringValue(SETTINGS, "CnCNetApiUrl", "http://cncnet-api/api/v1/");
```
Good catch!

CnCRAZER committed:
Clean UP

CnCRAZER committed:
Clean Up

CnCRAZER committed:
Fix typo in CnCNet API URL key

11EJDE11 commented on DXMainClient/DXGUI/Multiplayer/CnCNet/CnCNetAccountLoginPrompt.cs:
```diff
@@ -0,0 +1,90 @@
+            {
+                Name = "lblWindowTitle",
+                FontIndex = 1,
+                Text = "CONNECT TO CNCNET"
```
L10N all the visible-to-user strings for translations.

CnCRAZER committed:
Try to fix verification

Implement windows auth tokens alongside linux auth tokens for checkmark verification in nick lists

CnCRAZER committed:
Update CnCNetAPI.cs

SadPencil commented on DXMainClient/Domain/Multiplayer/CnCNet/CnCNetAPI.cs:
```diff
@@ -0,0 +1,364 @@
+                {
+                    var request = new NameValueCollection();
+                    request.Add("email", email);
+                    request.Add("password", password);
```
dude, HTTPS is considered safe, without mitm attacks unless the system ca is compromised

SadPencil replied:
all right...

Metadorius replied:
why even have the first hash?

SadPencil replied:
no. the 1st hash is calculated from the raw password. The server stores the 1st only to enable an HTTP-digest-like challenge, removing the need of transferring plaintext password as well as removing the need to store the plaintext password in the database

MahBoiDeveloper commented:
@CnCRAZER @SadPencil @SadPencil @Metadorius I find that FF's client ladder as lobbies with ladder icon in the regular lobby list (that neogrant's implement in old days) is a bit hacky instead of what we have right now in YR package -- separated button for quick matches in main menu. I suggest the future client's implementation of ladder API should follow the same idea -- separate window for searching quick matches in the ladder.

CnCRAZER committed:
Clean Up

CnCRAZER commented:
> Btw there should be button `Terms of Service` or something that redirects to CnCNet's ToS of ladder usage.

The register button takes you to the ladder account creation page. Not the exact same thing but meh

Metadorius commented:
> I find that FF's client ladder as lobbies with ladder icon in the regular lobby list (that neogrant's implement in old days) is a bit hacky instead of what we have right now in YR package -- separated button for quick matches in main menu. I suggest the future client's implementation of ladder API should follow the same idea -- separate window for searching quick matches in the ladder.

While I don't have much experience with FF's client ladder implementation, I disagree with the notion that the concept itself is bad.
1. There is *a ton* of ways to implement ranking and/or ladder, including the one using regular lobbies. Age of Empires II uses ranked custom lobbies which only get counted towards the ELO if certain conditions are met, for example, and it works fine.
2. Having to upkeep two separate clients using separate tech and implementations is a big maintenance headache. From my knowledge, only @Sneer-ra2 upkeeps the ladder client. What if he's gone? None of the XNA client devs could replace him properly.

CnCRAZER commented:
I wasn't expecting so much feedback so quickly lol.

So, currently: user has to go into an active ladder table and be sure they have an active username for the ladder to have the username show up in the login window. Am in talks with ladder api team to make a new _non ladder_ api for generic lobbies.


Will wait for that before I do any more changes I think

MahBoiDeveloper commented:
> Lets say a user decided to not use a cncnet account and use a nick of "xyz", but then a verified account logins with the same nick. Should the unverified account be kicked or told to change their name? Is this even possible? lol

I believe that provide users logging only with email+password would solve issue.

Currently if unregistered user would connect with the same name that already exist in the IRC chat, it would be renamed with `_` in the end, i.e. if I would make a fast reconnect to the Gamesurge IRC, my `mah_boi` nickname would stay in the net and new name would be `mah_boi_`.

If we talk about guests nicknames vs authorized nicknames, I prefer to kick guests or add them postfix like `mah_boi (Guest)` on logging and keep registered users without it.
